src/playground-project.ts

```diff
@@ -519,39 +520,76 @@ export class PlaygroundProject extends LitElement {
   }
 }
 
-const fetchProject = async (url: string) => {
-  const projectUrl = new URL(url, document.baseURI);
-  const manifestFetched = await fetch(url);
-  const manifest = (await manifestFetched.json()) as ProjectManifest;
-
-  const files = await Promise.all(
-    Object.entries(manifest.files ?? {}).map(
-      async ([name, {content, contentType, label, hidden}]) => {
-        if (content === undefined) {
-          const fileUrl = new URL(name, projectUrl);
-          const response = await fetch(fileUrl.href);
-          if (response.status === 404) {
-            throw new Error(`Could not find file ${name}`);
-          }
-          content = await response.text();
-          if (!contentType) {
-            contentType = response.headers.get('Content-Type') ?? undefined;
-          }
-        }
-        if (!contentType) {
-          contentType = typeFromFilename(name) ?? 'text/plain';
-        }
-        return {
-          name,
-          label,
-          hidden,
-          content,
-          contentType,
-        };
+const fetchProjectConfig = async (
+  url: string
+): Promise<{files: Array<SampleFile>; importMap: ModuleImportMap}> => {
+  const mergedFilePromises = new Map<string, Promise<FileOptions>>();
+  const mergedImportMap = {imports: {}};
+
+  const fetchAndMerge = async (configUrl: string) => {
```

Comment:
Can this closure be pulled out of the class or file and turned into a normal function? It doesn't seem to have element state side effects and passing in mergedFilePromises or mergedImportMap seems to be an okay solution.

My reasoning is that I think it's a bit less readable with it inlined like this. And to make it a bit more readable despite it mutating some of the arguments, you can make it return the same instance of `mergedFilePromises` and `mergedImportMap` in something like this

`{mergedFilePromises, mergedImportMap} = await fetchAndMerge(url, mergedFilePromises, mergedImportMap);`

Alternatively you can make it immutable but that's probably too much churn for "correctness"

Reply:
Good call, I refactored it into a single recursive function, hopefully better?